### PR TITLE
Add new option log_group_name_key

### PR DESCRIPTION
**Pull request by owlwalks**
*Issue #6 *

*Description of changes:*
Adding new option `log_group_name_key` and logics around it

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by PettitWesley** [comment]
@owlwalks Thank you for contributing this change. This and also setting the log stream based on a key is very useful IMO. 

Sorry it is taking us so long to get around to reviewing your code- we were busy with re:Invent. 

Can you rebase with the latest changes? We should get around to reviewing this soon.

**Comment by owlwalks** [comment]
hi! I just updated the PR, slightly different than the original one, it should be bit more manageable as logic of creating log group now stays in ```getLogStream()``` instead of ```createStream()```

File: cloudwatch/cloudwatch.go [hunk 1]
```diff
@@ -272,25 +272,36 @@ func (output *OutputPlugin) cleanUpExpiredLogStreams() {
...
 	stream, ok := output.streams[name]
-	if !ok {
-		// stream doesn't exist, create it
-		stream, err := output.createStream(tag)
+	if ok {
+		return stream, nil
+	}
+	// extra check, empty groupName indicates flush phase, but still land here meaning unable to find stream
+	if groupName == "" {
+		return nil, fmt.Errorf("error: does not compute: Log Stream %s could not be created, but also could not be found in the log group", name)
```
Comment (PettitWesley) [inline cloudwatch/cloudwatch.go]: This error message is copied from elsewhere, if there was ever a bug in this code it's more useful if it uses different text. May be something like "does not compute: in call to Flush() but log stream not found"

**Comment by PettitWesley** [comment]
@owlwalks First of all, let me again apologize for how long it has taken us to review this. Between re:invent, holiday vacations, and other Fluent Bit features, we've been very busy.

I like this feature; with it, some use cases are a lot more convenient. 

I see a couple of problems though with the current implementation:

1. If the log group key is not present, then the logs are rejected with an error. I understand the idea behind this, but I don't think its ideal. A lot of companies have a set up where one team manages infrastructure and deploys the Fluent Bit daemonset, and then other teams can run their application pods in a cluster. It's very easy for someone to accidentally misconfigure something and then lose all logs for one application because the log group key was incorrect. I'll ask my teammates if they agree on this- but I think it would be preferable if the `log_group_name` key is required with the `log_group_key` and serves as default/backup log group name. What do you think of that?

2. If I am reading the current code correctly, I think it will still always post to one log group: https://github.com/aws/amazon-cloudwatch-logs-for-fluent-bit/blob/0fbc162f52c97190e3c1c925df3cfe2206b6fa43/cloudwatch/cloudwatch.go#L524

I think that in order for this to work, we need to store the log group with the log stream- store both of them in the map of log streams (may be the key could be the concatenation of log group and log stream?).

3. Assuming I read the code correctly, I think that it will always create the log group? There does not seem to be any check to see if the log group already exists: https://github.com/aws/amazon-cloudwatch-logs-for-fluent-bit/blob/0fbc162f52c97190e3c1c925df3cfe2206b6fa43/cloudwatch/cloudwatch.go#L286

**Comment by kwizzn** [comment]
What does it take to get the ball rolling in this one? Having only one log group available is quite a blocker.

**Comment by PettitWesley** [comment]
@kwizzn This code needs to be re-written. If you'd like to contribute that- please do, it would be awesome. I suspect it'd be only a few days of work if you know your way around Golang and AWS.

**Comment by kwizzn** [comment]
OK, picked up @owlwalks's work in PR #49. Can be closed.

**Comment by hencrice** [comment]
Closing.

**Pull request closed without merging**